### PR TITLE
Combine x86 and i686 build scripts

### PR DESCRIPTION
**Pull request by stephengroat**
Let me know if you have any comments or changes.

Is there any reason that you haven't combine gcc and binutils? Or pushed this to homebrew? seems like something they would LOVE to have

### CODE REVIEW

**Comment by cosmo0920** [comment]
Hi, thanks for try to tackle for this.

> Is there any reason that you haven't combine gcc and binutils?

Because normal Windows users do not compile from source by themselves.
I want to use both of x86 and x64 versions of mingw toolchain at same time.
(20% of Windows users still use 32bit version of Windows.)

> Or pushed this to homebrew? seems like something they would LOVE to have

BTW, I think that this PR's formulae patches should apply into this repository when doing pull request Homebew mainline.

I have a question for this PR − Can your macOS environment compile mingw with gcc 5.0 or later?

**Comment by stephengroat** [comment]
to combine the `i686` and `x86_64`, look at the `binutils` script in c208ffc. i should be able to do something very similar in `gcc`

i haven't tried mingw with gcc 5.0 or later, something important i should try?

**Comment by cosmo0920** [comment]
> i haven't tried mingw with gcc 5.0 or later, something important i should try?

If you strongly want to send PR this formula into Homebrew mainline, you should try to build with latest gcc.

**Comment by cosmo0920** [comment]
And I enabled Travis CI in this repository.
Doesn't this formula excess build time limit in Travis CI?

**Review by cosmo0920 — CHANGES_REQUESTED** [review]
It looks current .travis.yml settings causes build timeout: https://travis-ci.org/stephengroat/homebrew-mingw_w64/builds/169852050

Could you check it?

**Comment by cosmo0920** [comment]
I've found the upgrading way to gcc 6.2.0.
Could you rebase at master and squash into a few commits?

**Comment by stephengroat** [comment]
i'm currently struggling trying to get `x86_64` and `i686` builds to run sequentially in the same directory

you can see the travis logs (https://travis-ci.org/stephengroat/homebrew-mingw_w64/builds/169967542)

any advice? i'm guessing that homebrew will want them to run directory

**Comment by stephengroat** [comment]
Currently take ~75 mins, seems to complete.

File: mingw-w64.rb [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        system "../mingw-w64-headers/configure", *args
+        system "make"
+        system "make", "install"
+      end
+
+      # TODO sgroat
+      #chdir install_prefix.to_s do
+      #  ln_s install_prefix.to_s, "#{libexec}/mingw"
+      #end
+
```
Comment (stephengroat) [inline mingw-w64.rb:71]: is removing this `ln` ok? is there a way to make add a `target_arch` to it? can't make both `i686` and `x86_64` and create the same `ln`

File: mingw-w64.rb [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        --host=#{target_arch}
+        --prefix=#{libexec}/#{target_arch}
+      ]
+
+      chdir "mingw-w64-libraries/winpthreads" do
+        system "./configure", *args
+        system "make"
+        system "make", "install-strip"
+      end
+
```
Comment (stephengroat) [inline mingw-w64.rb:194]: a little bit worried with what i say in the logs here. should the change in compiler be forcing a rebuild? should i run a `make clean` before configure?

```
==> ./configure CC=x86_64-w64-mingw32-gcc CXX=x86_64-w64-mingw32-g++ CPP=x86_64-w64-mingw32-cpp --host=x86_64-w64-mingw32 --prefix=/usr/local/Cellar/mingw-w64/4.0.6/libexec/x86_64-w64-mingw32
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for x86_64-w64-mingw32-strip... x86_64-w64-mingw32-strip
checking for a thread-safe mkdir -p... build-aux/install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking build system type... x86_64-apple-darwin16.0.0
checking host system type... x86_64-w64-mingw32
checking for x86_64-w64-mingw32-gcc... x86_64-w64-mingw32-gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... yes
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether x86_64-w64-mingw32-gcc accepts -g... yes
checking for x86_64-w64-mingw32-gcc option to accept ISO C89... none needed
checking whether x86_64-w64-mingw32-gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of x86_64-w64-mingw32-gcc... gcc3
checking for x86_64-w64-mingw32-ranlib... x86_64-w64-mingw32-ranlib
checking dependency style of x86_64-w64-mingw32-gcc... gcc3
checking for x86_64-w64-mingw32-ar... x86_64-w64-mingw32-ar
checking the archiver (x86_64-w64-mingw32-ar) interface... ar
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/local/Homebrew/Library/Homebrew/shims/super/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by x86_64-w64-mingw32-gcc... /usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-ld
checking if the linker (/usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-nm -B
checking the name lister (/usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 196608
checking how to convert x86_64-apple-darwin16.0.0 file names to x86_64-w64-mingw32 format... func_convert_file_nix_to_w32
checking how to convert x86_64-apple-darwin16.0.0 file names to toolchain format... func_convert_file_noop
checking for /usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-ld option to reload object files... -r
checking for x86_64-w64-mingw32-objdump... x86_64-w64-mingw32-objdump
checking how to recognize dependent libraries... file_magic ^x86 archive import|^x86 DLL
checking for x86_64-w64-mingw32-dlltool... x86_64-w64-mingw32-dlltool
checking how to associate runtime and link libraries... func_cygming_dll_for_implib
checking for x86_64-w64-mingw32-ar... (cached) x86_64-w64-mingw32-ar
checking for archiver @FILE support... @
checking for x86_64-w64-mingw32-strip... (cached) x86_64-w64-mingw32-strip
checking for x86_64-w64-mingw32-ranlib... (cached) x86_64-w64-mingw32-ranlib
checking command to parse /usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-nm -B output from x86_64-w64-mingw32-gcc object... ok
checking for sysroot... no
checking for a working dd... /bin/dd
checking how to truncate binary pipes... /bin/dd bs=4096 count=1
checking for x86_64-w64-mingw32-mt... no
checking for mt... no
checking if : is a manifest tool... no
checking how to run the C preprocessor... x86_64-w64-mingw32-cpp
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... no
checking for x86_64-w64-mingw32-as... x86_64-w64-mingw32-as
checking for x86_64-w64-mingw32-dlltool... (cached) x86_64-w64-mingw32-dlltool
checking for x86_64-w64-mingw32-objdump... (cached) x86_64-w64-mingw32-objdump
checking for objdir... .libs
checking if x86_64-w64-mingw32-gcc supports -fno-rtti -fno-exceptions... no
checking for x86_64-w64-mingw32-gcc option to produce PIC... -DDLL_EXPORT -DPIC
checking if x86_64-w64-mingw32-gcc PIC flag -DDLL_EXPORT -DPIC works... yes
checking if x86_64-w64-mingw32-gcc static flag -static works... yes
checking if x86_64-w64-mingw32-gcc supports -c -o file.o... yes
checking if x86_64-w64-mingw32-gcc supports -c -o file.o... (cached) yes
checking whether the x86_64-w64-mingw32-gcc linker (/usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-ld) supports shared libraries... yes
checking whether -lc should be explicitly linked in... yes
checking dynamic linker characteristics... Win32 ld.exe
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for x86_64-w64-mingw32-windres... x86_64-w64-mingw32-windres
checking limits.h usability... yes
checking limits.h presence... yes
checking for limits.h... yes
checking sys/timeb.h usability... yes
checking sys/timeb.h presence... yes
checking for sys/timeb.h... yes
checking for inline... inline
checking for size_t... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating tests/Makefile
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands
config.status: executing libtool commands
==> make
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-recursive
Making all in .
Making all in tests
make[2]: Nothing to be done for `all'.
```

File: mingw-w64.rb (hunk 1 above)
Comment (cosmo0920) [reply mingw-w64.rb:71]: Not OK.
This symbolic link is used in mingw compilation phase.

ref: https://sourceforge.net/p/mingw-w64/wiki2/Cross%20Win32%20and%20Win64%20compiler/

File: mingw-w64.rb (hunk 2 above)
Comment (cosmo0920) [reply mingw-w64.rb:194]: > should the change in compiler be forcing a rebuild? 

Yes. Could you use different directories in each architecture builds?

**Comment by cosmo0920** [comment]
This is just a my thought:

Combine into single formula improves maintainability?
If not, I will not be acceptable this change.

**Comment by cosmo0920** [comment]
> Currently take ~75 mins, seems to complete.

But, travis job limit will exceeded every time: https://travis-ci.org/cosmo0920/homebrew-mingw_w64/jobs/170004127

Travis job running time limit seems to be set up to 45min.

**Comment by stephengroat** [comment]
found the build timeouts: https://docs.travis-ci.com/user/customizing-the-build/#Build-Timeouts

> It is very common for test suites or build scripts to hang. Travis CI has specific time limits for each job, and will stop the build and and add an error message to the build log in the following situations:
> 
> A job takes longer than 50 minutes on travis-ci.org
> A job takes longer than 120 minutes on travis-ci.com
> A job takes longer than 50 minutes on OSX infrastructure or travis-ci.org or travis-ci.com
> A job produces no log output for 10 minutes

**Comment by cosmo0920** [comment]
I don't want to concatenate into all-in-one formula.
Because my macOS machine does not have enough machine power to build it within reasonable time (~50min).

Could you send these patches into Homebrew mainline instead of this repository?
It is preferable for me.

**Comment by stephengroat** [comment]
sure. hopefully they'll get a keg build going so individual builds aren't
necessary

**Comment by stephengroat** [comment]
will look into the newer mingw builds

you can cheat with travis, travis_wait can be set higher (90 minutes seems
to be ok) or you can only compile x86_64, which takes ~30 mins

**Comment by cosmo0920** [comment]
> you can cheat with travis, travis_wait can be set higher (90 minutes seems
> to be ok) or you can only compile x86_64, which takes ~30 mins

But I couldn't succeeded to build within time limit in Travis.
Because Travis CI environment is slow and have to install dependent Formulae in every time.

**Comment by stephengroat** [comment]
@cosmo0920 I know that i closed this pr, but i've still been working on the script and I think I'm super close. any idea with this error message? is the program `ld` not finding `libgcc.a` or is `hello-c.c` not finding `libgcc.a`?

```
==> x86_64-w64-mingw32-gcc -o hello-c hello-c.c
Last 15 lines from /Users/sgroat/Library/Logs/Homebrew/mingw-w64/test.01.x86_64-w64-mingw32-gcc:
2016-10-26 13:39:30 -0700

x86_64-w64-mingw32-gcc
-o
hello-c
hello-c.c

/usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-ld: cannot find -lgcc
/usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-ld: cannot find -lgcc_eh
/usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-ld: cannot find -lgcc
/usr/local/Cellar/mingw-w64/4.0.6/libexec/bin/x86_64-w64-mingw32-ld: cannot find -lgcc_eh
collect2: error: ld returned 1 exit status
```

**Pull request closed without merging**